### DISCUSSION TITLE
Poiji XXE prevent attack

### DISCUSSION THREAD

**Discussion by ruipbferreira** [ozlerhakan/poiji#264] [Q&A] [answered]
Hi all,
I'm using Poiji to process excel files and I have a question related to XXE attacks. How can I prevent it in Poiji? 

Thank you all for your help
RF

**Comment by ozlerhakan** [comment, marked as answer]
Hi @ruipbferreira ,

Digging into XXE attacks a bit, I have found that the current Poiji is not affected by this as we use XMLBeans whose version is greater than 4.0.0 according :

> 13 January 2021 - CVE-2021-23926 - XML External Entity (XXE) Processing in Apache XMLBeans versions prior to 3.0.0
  Description:
  When parsing XML files using XMLBeans 2.6.0 or below, the underlying parser created by XMLBeans could be susceptible to XML 
 External Entity (XXE) attacks.
  This issue was fixed a few years ago but on review, we decided we should have a CVE to raise awareness of the issue.
  Mitigation:
  Affected users are advised to update to Apache XMLBeans 3.0.0 or above which fixes this vulnerability. XMLBeans 4.0.0 or above is preferable.

**Reply by ruipbferreira** [reply]
Hi @ozlerhakan thanks for your response. I'm using poiji 3.0.3 and I can perform XXE attack. I'll update to latest version and retest

**Reply by ruipbferreira** [reply]
@ozlerhakan I had the same result :/

**Reply by ozlerhakan** [reply]
Why do you use 3.0.3?

**Reply by ozlerhakan** [reply]
Poiji 3.0.3 uses 3.1.0 XMLBeans underneath.

**Reply by ozlerhakan** [reply]
Could you please provide your attempts or use cases for this scenario?

**Reply by ozlerhakan** [reply]
Make sure that the relevant transitive dependencies are loaded correctly.

**Reply by ruipbferreira** [reply]
@ozlerhakan i use 3.0.3 because of legacy. I will try again with the new one. Can you please clarify witch decencies i should validate?

**Reply by ozlerhakan** [reply]
once you load all the dependencies, be sure that Poiji makes use of  xmlbeans-x.x.x.jar where X.X.X should be 4.0.0 or above.

**Reply by ruipbferreira** [reply]
@ozlerhakan version 3.1.9 works with java 8? I have the following compile error  
\com\github\ozlerhakan\poiji\3.1.9\poiji-3.1.9.jar(com/poiji/annotation/ExcelCellName.class
class file has wrong version 61.0, should be 52.0

**Reply by ozlerhakan** [reply]
You could try other versions that don't need Java 17 VM. The next release will run in Java 11 back again.

**Reply by ruipbferreira** [reply]
@ozlerhakan  I will try 3.1.1 and ensure the xmlbeans version. Just to clarify, the poiji have dependency to xmlbeans?

**Reply by ruipbferreira** [reply]
@ozlerhakan attached the dependency tree. With this I still can perform XXE. Attached a video also



![Dependency](https://user-images.githubusercontent.com/17995212/209834665-b15d3939-15a5-43ec-ab6d-201742c5cb0d.png)

**Reply by ozlerhakan** [reply]
Apache Poi uses XMLBeans as a dependency by itself. Interesting that they mention this attack in their home page https://poi.apache.org/, and the vulnerability is said to be fixed with 4.0.0 and above versions.

**Reply by ozlerhakan** [reply]
I'll make a new snapshot of 3.2.0, so that you can give a try again.

**Reply by ruipbferreira** [reply]
@ozlerhakan im trying to update Apache-poi to 5.x.x version. I think it will solve it

**Reply by ozlerhakan** [reply]
this is my main reason to upgrade poiji as well too.

**Reply by ozlerhakan** [reply]
could you please try 3.2.0 that I have just released @ruipbferreira ? It should use 5.1.1 XMLBeans.

**Reply by ruipbferreira** [reply]
@ozlerhakan i Will try. Let you know the result

**Reply by ruipbferreira** [reply]
@ozlerhakan where I an get 3.2.0? I check here https://mvnrepository.com/artifact/com.github.ozlerhakan/poiji

**Reply by ozlerhakan** [reply]
Here it is: https://central.sonatype.dev/artifact/com.github.ozlerhakan/poiji/3.2.0

**Reply by ruipbferreira** [reply]
@ozlerhakan I found it and I'm unable to test it "class file has wrong version 55.0, should be 52.0" I'm working with java 8

**Reply by ozlerhakan** [reply]
Ah, Poiji uses Java 11.

**Reply by ruipbferreira** [reply]
@ozlerhakan can you tell me the last version with java 8?

**Reply by ruipbferreira** [reply]
@ozlerhakan I've update poi-ooxml to 5.2.3 and have the error parsing excel  "DOCTYPE is disallowed when the feature \"http://apache.org/xml/features/disallow-doctype-decl\" set to true.". To do that I excluded poi-ooxml dependency from poiji 3.0.3. I will update my Poiji dependency to the last one supported in java 8. Thanks for your help and keep going with your good work

**Reply by ozlerhakan** [reply]
Great! I was going to mention this method 😎 thank you for using Poiji. You can mark this thread as complete.
